dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/engine/DefaultProgramRuleEngine.java

```diff
@@ -148,12 +122,44 @@ public RuleValidationResult getDataExpressionDescription(String dataExpression,
             constantService.getAllConstants()));
   }
 
+  @Override
+  public RuleEngineEffects evaluateEnrollmentAndTrackerEvents(
+      @Nonnull RuleEnrollment enrollment,
+      @Nonnull List<RuleEvent> events,
+      @Nonnull Program program,
+      @Nonnull UserDetails user) {
+    return evaluateEnrollmentAndEvents(enrollment, events, program, user);
+  }
+
+  @Override
+  public RuleEngineEffects evaluateProgramEvents(
+      @Nonnull List<RuleEvent> events, @Nonnull Program program, @Nonnull UserDetails user) {
+    return evaluateEnrollmentAndEvents(null, events, program, user);
+  }
+
+  private RuleEngineEffects evaluateEnrollmentAndEvents(
+      RuleEnrollment enrollment,
```

Comment:
`@CheckForNull` is the counterpart to `@Nonnull`. I don't think we need to add these to all methods public ones should be enough 🤔 the analyzer should then be able to analyze the different code paths of private methods as well.

Reply:
To be honest, here I thought about it, but I was not sure whether to add it.
Maybe it is just the naming that is confusing me, but I am not going to check the null in the method, I am just passing the value through, no matter if it is null or not. Does it still make sense to annotate with @CheckForNull?
I mean, if the annotation was called @Nullable, I wouldn't have any doubt.
I am going to add it anyway, the analyzer is going to be happy and this is what matters.